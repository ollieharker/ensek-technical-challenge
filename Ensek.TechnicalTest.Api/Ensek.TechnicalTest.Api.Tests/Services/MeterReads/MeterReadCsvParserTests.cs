using ApprovalTests;
using ApprovalTests.Reporters;
using Ensek.TechnicalTest.Api.Exceptions;
using Ensek.TechnicalTest.Api.Services.MeterReads;
using Newtonsoft.Json;
using System.Text;

namespace Ensek.TechnicalTest.Api.Tests.Services.MeterReads
{
	[Trait("IO", "CSV")]
	[UseReporter(typeof(DiffReporter))]
	public class MeterReadCsvParserTests
	{
		private CsvHelperMeterReadCsvParser sut;

		public MeterReadCsvParserTests()
		{
			this.sut = new CsvHelperMeterReadCsvParser();
		}

		[Fact]
		public void GivenNullStreamExceptionThrown()
		{
			Assert.Throws<ArgumentNullException>(() => this.sut.ParseCsvToMeterReadings(null));
		}

		[Fact]
		public void GivenEmptyStreamNoRecordsReturned()
		{
			var result = this.sut.ParseCsvToMeterReadings(new MemoryStream());

			Assert.Empty(result.NewMeterReadings);
		}

		[Fact]
		public void GivenEmptyStreamNoFailedToParseRecords()
		{
			const int expectedFailedParse = 0;

			var result = this.sut.ParseCsvToMeterReadings(new MemoryStream());

			Assert.Equal(expectedFailedParse, result.ParseFailedCount);
		}

		[Fact]
		public void GivenNonCsvTextMeterParseExceptionThrown()
		{
			var bytes = Encoding.UTF8.GetBytes(Guid.NewGuid().ToString());
			var memoryStream = new MemoryStream(bytes);

			Assert.Throws<MeterReadParsingException>(() => this.sut.ParseCsvToMeterReadings(memoryStream));
		}

		[Fact]
		public void GivenValidCsvRecordsAddedResult()
		{
			const string path = "./Data/MeterReadings/valid-readings.csv";
			this.ParseAndApproveTestCsvFile(path);
		}

		[Fact]
		public void GivenInvalidCsvRecordsBadResultsRemoved()
		{
			const string path = "./Data/MeterReadings/invalid-readings.csv";
			this.ParseAndApproveTestCsvFile(path);
		}

		[Fact]
		public void GivenInvalidAndValidCsvRecordsBadResultsRemoved()
		{
			const string path = "./Data/MeterReadings/readings-mixed-validity.csv";
			this.ParseAndApproveTestCsvFile(path);
		}

		private void ParseAndApproveTestCsvFile(string filePath)
		{
			var fileString = this.LoadCsvFileIntoStream(filePath);

			var result = this.sut.ParseCsvToMeterReadings(fileString);

			Approvals.Verify(JsonConvert.SerializeObject(result, Formatting.Indented));
		}

		private Stream LoadCsvFileIntoStream(string filePath)
		{
			return File.OpenRead(filePath);
		} 
	}
}
