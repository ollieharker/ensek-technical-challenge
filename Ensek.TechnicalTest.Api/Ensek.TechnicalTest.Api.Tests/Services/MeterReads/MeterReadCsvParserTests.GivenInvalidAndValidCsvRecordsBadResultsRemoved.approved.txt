{
  "ParseFailedCount": 4,
  "NewMeterReadings": [
    {
      "Id": 0,
      "Account": null,
      "AccountId": 1,
      "DateTime": "2019-04-22T09:24:00+01:00",
      "Value": 343
    },
    {
      "Id": 0,
      "Account": null,
      "AccountId": 1000,
      "DateTime": "2019-04-22T09:24:00+01:00",
      "Value": 5322
    },
    {
      "Id": 0,
      "Account": null,
      "AccountId": -1,
      "DateTime": "2019-04-22T09:24:00+01:00",
      "Value": 3
    }
  ]
}